build(client-tree): exports with ESM as primary

jason-ha opened the pull request:
- Configure full ESM build and declaration for tree
- Set exports (only "public" and `internal` as there are no others) 
  - "@internal" uses of tree updated to import "/internal"
- Build CommonJS secondarily with injected CommonJS package.json at build

A module package.json is also injected into ESM build for symmetry.

ESM api-extractor tsconfig is overwritten to use `Bundler` resolution as `Node16` does not appear to be robustly handled.

jason-ha committed:
build: tsc-multi for cjs renames .d.ts

jason-ha committed:
build(tree): settings for ESM as primary
- exports not specified
- API extractor for CJS does not work properly

jason-ha committed:
build(client-tree): CJS via tsc

jason-ha committed:
test(client-tree): test commonjs

jason-ha committed:
build(client-tree): exclude fluid-build-tasks-tsc policy

jason-ha committed:
build(client-tree): set package exports

jason-ha commented on packages/dds/tree/package.json:
```diff
@@ -127,6 +151,34 @@
+			"ci:build:docs": [
+				"build:esnext"
+			],
+			"commonjs": {
```
I don't think the dependency changes here and below are needed. Look like leftovers when there was an attempt to have `tsc` be for ESM build.

jason-ha committed:
build(client-tree): cleanup build dependencies

jason-ha committed:
build(client-tree): full ESM build

jason-ha committed:
build(client-tree): add api task
remove unused build:gen-version

jason-ha committed:
build(client-tree): fix api report/rollup
make sure cjs rollup uses cjs config
override moduleResolution for esm as api-extractor doesn't handle node16 robustly. It will generate with imports to invalid paths.

jason-ha commented on packages/dds/tree/api-extractor.json:
```diff
@@ -1,4 +1,15 @@
+			"$schema": "http://json.schemastore.org/tsconfig",
+			"extends": "./tsconfig.esnext.json",
+			"compilerOptions": {
+				"module": "esnext",
+				"moduleResolution": "node"
```
Yes, that is better. I was trying to pick whatever was happening before changing tsconfig to moduleResolution=Node16. In before case for esnext build, module=esnext and moduleResolution=node per common/build/build-common/ts-common-config.json. For API extractor, before, it was module=commonjs based; so, keeping node wasn't fully keeping things like before. This isn't generating traditional output and I don't think is impacted by that TS concern, but I agree using approved pairing is best.

jason-ha committed:
build(client-tree): improve api report/rollup
use Bundler for moduleResolution with ESNext

tylerbutler commented on common/build/build-common/tsc-multi.cjs.json:
```diff
@@ -1,4 +1,4 @@
 {
-	"targets": [{ "extname": ".cjs", "module": "node16" }],
+	"targets": [{ "extname": ".cjs", "dExtName": ".d.cts", "module": "node16", "outDir": "./dist" }],
```
Nit: you may want to make this change separately since it will affect more than just tree.

jason-ha replied:
reverted - I no longer think we should use tsc-multi; so, I'll just drop it.

tylerbutler commented on packages/dds/tree/package.json:
```diff
@@ -11,18 +11,42 @@
+	"exports": {
+		".": {
+			"import": {
+				"types": "./lib/tree-public.d.ts",
```
Suggest making this change separately.

jason-ha replied:
Are you sure? We want this make sure all is good downstream with the exports here. Having this helped me validate (and then fix things).

tylerbutler commented on packages/dds/tree/tsconfig.base.json:
```diff
@@ -0,0 +1,17 @@
+{
+	"extends": "@fluidframework/build-common/ts-common-config.json",
```
Any reason not to use the new shared configs here?

jason-ha replied:
When I look in @fluidframework/build-common that is the only config file. Workspace needs an @fluidframework/build-common update?

tylerbutler replied:
We used relative paths so we don't have to publish build-common all the time. Look at merge-tree or other packages for the new pattern. Very open to changes to the pattern in the future if you have feedback on it.

See also https://github.com/microsoft/FluidFramework/pull/18152 but the pr description is not completely accurate. Especially about node10 and esnext. Oh, I was so naive then :)

jason-ha replied:
I've seen those and thought sure that must be temporary. That is far too many `../`. Is the plan not to publish then?
We can do a bit better. For now I would go with getting the settings right (as I have them). And then we can roll over to a nicer pattern of common definitions.
Won't fix for this PR.

tylerbutler replied:
We can eventually publish but IMO relative paths are a lot better for in-repo infrastructure because it allows for instant application of changes throughout the release group. If we had not used this pattern when making these changes we would have had to publish at least three versions of the package.

We can still publish the package if/when things are settled and/or we want to make the changes more easily available to other release groups.

jason-ha replied:
Yes, having a way to consume instantly is good. I don't think we need to publish to resolve the general relative path issue. But I still think not for this change. A nice thing to do after; here keep changes and no changes very clear.

jason-ha committed:
revert(build-common): don't update tsc-multi.cjs.json

jason-ha committed:
merge: main with tree-exports-with-esm-as-primary

jason-ha committed:
merge: main with tree-exports-with-esm-as-primary

jason-ha committed:
fix(tree): fix CJS package.json injection

echo quote emit syntax is different per shell.
Instead copy a file.

jason-ha committed:
build(client-tree): CJS explicit moduleResolution
Add comments for module/moduleResolution settings
and make moduleResolution explicit.
Also inject package.json for ESM for symmetry.

tylerbutler commented:
I dug up the comment that describes some drawbacks to this approach:

https://github.com/microsoft/TypeScript/issues/54593#issuecomment-1585196329

I don't think this is an immediate blocker for us, but it makes me think this is a design that we'll still have to revisit later. Maybe just something we note in the readme or dev notes and revisit when it becomes a practical blocker.

jason-ha commented:
> I dug up the comment that describes some drawbacks to this approach:
> 
> [microsoft/TypeScript#54593 (comment)](https://github.com/microsoft/TypeScript/issues/54593#issuecomment-1585196329)
> 
> I don't think this is an immediate blocker for us, but it makes me think this is a design that we'll still have to revisit later. Maybe just something we note in the readme or dev notes and revisit when it becomes a practical blocker.

I was emitting `lib/package.json` for complete symmetry and we can do that or not. I think nice to have for now to help catch any issues from this structure. We don't use "imports" feature and I don't see the need to; so, we can add a policy to block. As noted in the referenced thread it is possible to enhance those injected package.json file with other information if needed.

I am hoping that we are really working toward an ESM only future and would then be able to drop secondary CJS build. If we can't before there is issue with `lib/package.json` then we will have option to drop it with current configuration.

Those seem more tractable than using tools that we can't rely on as much as tsc.

Last weekend I was playing around with addition to tsc-multi to hook package.json reads and override type. Using it for just that feature seems safe to me. (No extension or file rewrites.)

tylerbutler commented on packages/dds/tree/api-extractor.json:
```diff
@@ -1,4 +1,15 @@
 {
 	"$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
-	"extends": "../../../common/build/build-common/api-extractor-base.json"
+	"extends": "../../../common/build/build-common/api-extractor-base.json",
```
Is this needed? Seems like the package.json scripts explicitly pass a config file.

jason-ha replied:
- Everything eventually extends from api-extractor-base. The other build-common files extend api-extractor-base, set rollup files, and if not primary disable apiReport.
- `build:docs` and `ci:build:docs` use this file implicitly.

tylerbutler commented on packages/dds/tree/package.json:
```diff
@@ -11,18 +11,42 @@
 	"license": "MIT",
 	"author": "Microsoft and contributors",
 	"sideEffects": false,
+	"type": "module",
```
Isn't this a bit dangerous?

jason-ha replied:
I don't know that this is dangerous. Why might it be dangerous?

jason-ha replied:
(I haven't requested the tsc-multi feature that I think it needs. Without tsc seeing "type": "module" it doesn't go into full ESM mode. This is a must for now to ensure proper ESM transpilation.)

tylerbutler commented on packages/dds/tree/package.json:
```diff
@@ -46,7 +70,7 @@
 		"test:stress": "cross-env FUZZ_TEST_COUNT=20 FUZZ_STRESS_RUN=true mocha --ignore \"dist/test/memory/**/*\" --recursive \"dist/test/**/*.spec.js\" -r @fluidframework/mocha-test-setup",
 		"ts2esm": "ts2esm ./tsconfig.json ./src/test/tsconfig.json",
-		"tsc": "tsc",
+		"tsc": "tsc --project ./tsconfig.cjs.json && copyfiles -f ../../../common/build/build-common/src/cjs/package.json ./dist",
```
I think this is fine for now since fluid-build understands tsc and copyfiles and should decompose this into two dependent tasks (one for tsc and the other for copyfiles). However, longer-term I think the pattern we should use is to have explicit separate tasks for the copying and make those dependencies explicit in the fluid-build config as well.

jason-ha replied:
I did enjoy fluid-build being smart about this. I am not sure the long-term desired breakup. I guess `tsc` would kick off the two tasks b/c we want to keep the main build command simple and common. So, we have tsc that does tsc:transpile and tsc:setcjs later?

tylerbutler commented on packages/dds/tree/tsconfig.cjs.json:
```diff
@@ -0,0 +1,17 @@
+		// can be Node16 with a "type": "commonjs" package.json override.
+		// Underlying reason for the complexity here is that base tools wish to avoid the dual
+		// package hazard of ESM and CommonJS in the same package. "tsc-multi" is a workaround.
+		"module": "CommonJS",
+		"moduleResolution": "Node10",
```
Are we certain this produces valid output?

jason-ha replied:
Yes. This is not a change from what tree was setup for before. moduleResolution was node which is the deprecated specification for Node10.

jason-ha committed:
merge: main with tree-exports-with-esm-as-primary

jason-ha commented on packages/dds/tree/api-extractor-cjs.json:
```diff
@@ -0,0 +1,10 @@
+{
+	"$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
+	"extends": "../../../common/build/build-common/api-extractor-base.cjs.primary.json",
+	"compiler": {
+		"tsconfigFilePath": "<projectFolder>/tsconfig.cjs.json"
+	},
+	"apiReport": {
+		"enabled": false
+	}
```
I am not sure what the question is.
- Should tsconfigFilePath be set in build-common/api-extractor-base.cjs*.json? No each package may have its own tsconfig setup at the moment.
- Should apiReport be disabled in build-common/api-extractor-base.json? I think it should probably be off in there. The -base.*.primary.json files however would want it on. This config wants the extension neutral (.ts) settings from .primary.json but not the apiReport.

jason-ha commented on packages/dds/tree/package.json:
```diff
@@ -11,18 +11,42 @@
 		"build:compile": "fluid-build . --task compile",
 		"build:docs": "api-extractor run --local",
-		"build:esnext": "tsc --project ./tsconfig.esnext.json",
-		"build:genver": "gen-version",
```
So long ago - perhaps bad merge?

jason-ha committed:
build(client-tree): restore build:genver

jason-ha committed:
merge: main with tree-exports-with-esm-as-primary

jason-ha committed:
Merge branch 'main' into tree-exports-with-esm-as-primary

jason-ha commented:
Ah timing. Looks like some non-public use of tree snuck in. exports will need adjusted (if not removed).

jason-ha committed:
Merge branch 'main' into tree-exports-with-esm-as-primary

jason-ha committed:
build(client): fix internal tree imports

alexvy86 commented on packages/tools/devtools/devtools-core/src/data-visualization/DefaultVisualizers.ts:
```diff
@@ -13,7 +13,8 @@ import { SharedCounter } from "@fluidframework/counter";
 import { SharedMatrix } from "@fluidframework/matrix";
 import { SharedString } from "@fluidframework/sequence";
-import { SharedTree, type ISharedTree, encodeTreeSchema } from "@fluidframework/tree";
+import type { ISharedTree } from "@fluidframework/tree/internal";
+import { SharedTree, encodeTreeSchema } from "@fluidframework/tree/internal";
```
Just surprised to see these are all internal now. Are we sure not even ISharedTree should be public?

alexvy86 reviewed (approved):
Just looked at the Devtools change in detail. Can't vouch for the rest :)
